### PR TITLE
8 dna interpreter experiment alex

### PR DESCRIPTION
**Pull request by alexmorten**
https://trello.com/c/F6GLeUMn/34-simple-dna-interpreter-experiment
Includes https://github.com/codeuniversity/al-cis/pull/7

This is not perfect yet, but I think this is a good basis for discussion

### CODE REVIEW

File: dna_decoding.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        current_connection_count,
+    )
+    max_start_index = len(dna) - 1
+    start_index = round(start_feature_value * max_start_index)
+    max_end_step = (len(dna) - 1) - start_index
+    end_index = start_index + round(max_end_step * end_feature_value) + 1
+    return dna[start_index:end_index]
+
+
+def shift_bits_by(bits, n):
```
Comment (Animuslupus) [inline dna_decoding.py]: Why didn't you used the build-in bitwise shift operators `>>` (right shift) and `<<` (left shift)?

File: cis_env.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        conf.WORLD_DIMENSION[1] / 2,
+        conf.WORLD_DIMENSION[2] / 2
+    ]),
+    0.002
+)
+
+
+def feed_cell(cell, cell_dict, already_averaged_dict, food_factor=1):
+    f = random.uniform(0, 1)
+    if f < dna_decoding.food_theshold(cell.dna):
```
Comment (Animuslupus) [inline cis_env.py]: Doesn't the food threshold decode the probability for a cell to find food. Why should it be part of the cells DNA?

File: dna_decoding.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    if random_bool_with_threshold(chance):
+        new_dna_bit_string += '0' * 8
+    return BitArray(bin=new_dna_bit_string).bytes
+
+
+def random_bool_with_threshold(threshold):
+    return random.uniform(0, 1) < threshold
+
+
+def mutate_bit_string(bit_string):
```
Comment (Animuslupus) [inline dna_decoding.py]: Normally a mutation in genetic algorithm switches an existing bit ([Genetic Algorithm - Mutation](https://en.wikipedia.org/wiki/Mutation_(genetic_algorithm))). Why did you choose to not allow bit-switching, but only to add new bits instead?

File: cis_env.py (hunk 2 above)
Comment (alexmorten) [reply cis_env.py]: It shouldn't be 😅 
I just used that to make sure this works as expected

Comment (alexmorten) [reply cis_env.py]: I would suggest to swap that out as soon as we have the food function and proper interactions between cells that are not connected

**Pull request merged**